langkit/templates/pkg_introspection_body_ada.mako

```diff
@@ -268,27 +346,94 @@ package body ${ada_lib_name}.Introspection is
    function Fields (Kind : ${root_node_kind_name}) return Field_Reference_Array
    is
    begin
-      % if ctx.generic_list_type.concrete_subclasses:
-         if Kind in ${ctx.generic_list_type.ada_kind_range_name} then
-            return (1 .. 0 => <>);
-         end if;
+      % if ctx.sorted_parse_fields:
+         return Fields (Id_For_Kind (Kind), Concrete_Only => True);
+      % else:
+         ${return_program_error()}
       % endif
+   end Fields;
 
+   ------------
+   -- Fields --
+   ------------
+
+   function Fields
```

Comment:
This looks like expensive work that could be cached no ?

Reply:
As discussed live: this algorithm runs in linear time for number of fields + number of derivations. Caching is possible, but I would suggest to wait for reported performance issue to bother introducing a cache. :-)